sprinkle/low-level/context.py

```diff
@@ -0,0 +1,127 @@
+context.rendezvous.get_node_rank()
+context.rendezvous.get_slots_per_node()
+
+# master data: applies to any job running on-cluster, of any job that you have
+# chosen to connect to the master with context.connect_to_master()
+context.connect_to_master(...)  # args similar to Determined() object
+context.master.get_addr()
+context.master.get_port()
+context.master.get_tls()
+
+# cluster data: applies to any job running on-cluster
+context.cluster.get_container_id()
+context.cluster.get_container_gpus()
+context.cluster.get_slot_ids()
+
+# training-related data; only applies to `det e create` workloads
+context.training.get_experiment_config()
+context.training.get_data_config()
+context.training.get_experiment_id()
+context.training.get_trial_id()
+context.training.get_global_batch_size()
+context.training.get_per_slot_batch_size()
+context.training.get_hparams()
+context.training.get_trial_seed()
+context.training.get_latest_checkpoint()
+
+"""
+PUSH-ARCHITECTURE RELATED APIS:
+
+The core features of our platform can be delivered if we only have metrics and checkpoints for
+training.  Advanced features like preemption and hpsearch should also be possible, but MUST NOT
+be necessary to deliver basic features.
+"""
+
+# Training metrics
+context.training.begin_training()  # optional call, improves webui experience
+context.training.report_training_metrics(
+    metrics=...         # optional: reduced metrics, shown in webui
+    batch_metrics=...   # optional: accessible via python sdk
+    batches_trained=... # optional: epochs/batches/shows in webui
+    records_trained=... # optional: shows in webui
+    start_time=...      # optional: shows in webui
+    end_time=...        # optional: shows in webui
+)
+
+# Validation metrics
+context.training.begin_validation()  # optional call, improves webui experience
+context.training.report_validation_metrics(
+    metrics=...       # required: reduced metrics, shown in webui
+    start_time=...    # optional: shows in webui
+    end_time=...      # optional: shows in webui
+)
+
+# Checkpoints
+context.api._begin_checkpoint()                             # non-user-facing call
+context.api._report_checkpoint(uuid, start_time, end_time)  # non-user-facing call
+# user-facing API, just wraps the StorageManagers:
+with context.checkpoint.save_path as path:
+    ... # user saves checkpoint into path
+# TODO: for downloading, checkpoints, do we just stick to some form of Checkpoint Export API?
+
+
+# Searcher API:
+# (fault tolerant via context.training.get_latest_checkpoint)
+op = context.training.get_searcher_op()
+    # op will be one of TrainingOp() or ValidationOp()
+training_op.complete(
+    checkpoint=...    # required: an already-completed checkpoint uuid
+)
+validation_op.complete(
+    searcher_metric=...  # required
+)
+
+
+# Preemption API:
+# (different than adaptive's  early stopping; more like the cancel button or a spot instance)
+context.distributed.should_preempt(period=10)
```

Comment:
I struggled a bit with where to place it.  But I've been thinking the `context.distributed` object should really have a zmq-based `allgather()` method, and that method (or the asynchronous version of it), would be used to do the preemption.  If you put the preemtion anywhere else, it would need access to the `context.distributed` object, which would be slightly messy.

Reply:
> If you put the preemtion anywhere else, it would need access to the context.distributed object, which would be slightly messy.

I think there is no need to worry about it since they are all internal. You can put the class definition for `context.distributed` as an inner class.

Anyways, `should_preempt` is user-facing. It would be weird to put it in `distributed`.

Reply:
Another reason is that if we support spot instances, this API might be able to also handle that. In this case, synchronizing between containers might not be useful